honeybee/aperture.py

```diff
@@ -0,0 +1,506 @@
+            if raise_exception:
+                raise ValueError('Aperture "{}" has self-intersecting edges.'.format(
+                    self.name_original))
+            return False
+        return True
+
+    def is_non_zero(self, tolerance=0.0001, raise_exception=True):
+        """Check whether the area of the Aperture is above a certain "zero" tolerance.
+
+        Args:
+            tolerance: The minimum acceptable area of the object. Default is 0.0001,
+                which is equal to 1 cm2 when model units are meters. This is just
+                above the smalest size that OpenStudio will accept.
+            raise_exception: If True, a ValueError will be raised if the object
+                area is below the tolerance. Default: True.
+        """
+        if self.area < tolerance:
+            if raise_exception:
+                raise ValueError(
+                    'Aperture "{}" geometry is too small. Area must be at least {}. '
+                    'Got {}.'.format(self.name_original, tolerance, self.area))
+            return False
+        return True
+
+    @property
+    def properties(self):
+        """Aperture properties, including Radiance, Energy and other properties."""
+        return self._properties
+
+    @property
+    def to(self):
+        """Aperture writer object.
+
+        Use this method to access Writer class to write the aperture in other formats.
+
+        Usage:
+            aperture.to.idf() -> idf string.
+            aperture.to.radiance() -> Radiance string.
+        """
+        return self._writer
+
+    def to_dict(self, included_prop=None):
+        """Return Aperture as a dictionary.
+
+        Args:
+            included_prop: List of properties to filter keys that must be included in
+                output dictionary. For example ['energy'] will include 'energy' key if
+                available in properties to_dict. By default all the keys will be
+                included. To exclude all the keys from plugins use an empty list.
+        """
+        base = {
+            'type': self.__class__.__name__,
+            'name': self.name,
+            'name_original': self.name_original,
+            'geometry': self._geometry.to_dict(),
+            'aperture_type': self.type.to_dict(),
+            'properties': self.properties.to_dict(included_prop)
+        }
+        if 'energy' in base['properties']:
+            base['boundary_condition'] = self.boundary_condition.to_dict(full=True)
+        else:
+            base['boundary_condition'] = self.boundary_condition.to_dict(full=False)
+        if self.parent:
+            base['parent'] = self.parent.name
+        return base
+
+    def duplicate(self):
+        """Get a copy of this object."""
+        return self.__copy__()
+
+    def ToString(self):
+        return self.__repr__()
+
+    def __copy__(self):
+        new_ap = Aperture(self.name_original, self.geometry, self.type.duplicate(),
```

Comment:
I might be wrong but I think we should not need to duplicate type and boundary condition except for Surface that boundary condition should be duplicated.

Reply:
I absolutely agree here.  Now that the `type` and `boundary_condition` are immutable objects, there's no need to duplicate them.  So I removed the `duplicate` methods from them along with the code here.